scripts/gen_ldelf_hex.py

```diff
@@ -48,16 +50,25 @@ def emit_load_segments(elffile, outf):
                     sys.exit(1)
                 code_size = segment['p_filesz']
             else:
-                if segment['p_flags'] != (P_FLAGS.PF_R | P_FLAGS.PF_W):
-                    print('Expected load segment to be read/write')
-                    sys.exit(1)
-                if next_rwseg_va and segment['p_vaddr'] != next_rwseg_va:
-                    print('Expected contiguous read/write segments')
-                    print(segment['p_vaddr'])
-                    print(next_rwseg_va)
-                    sys.exit(1)
-                data_size += segment['p_filesz']
-                next_rwseg_va = segment['p_vaddr'] + segment['p_filesz']
+                if segment['p_flags'] & P_FLAGS.PF_W != P_FLAGS.PF_W:
+                    # A subsequent non-writable segment may be placed in the
+                    # "code" area (mapped RX)
+                    if data_size != 0:
+                        # Already met RW data: error
+                        print(f'Expected writable load segment (n={n})')
+                        sys.exit(1)
+                    else:
+                        if next_rseg_va and segment['p_vaddr'] != next_rseg_va:
```

Comment:
This doesn't catch a gap between the RX and this R segment. Perhaps we can assign expected value above when taking care of the RX segment?

Reply:
Good catch, I'll fix that

Reply:
Hmmm... the problem is, Clang may also generate a gap ;-/ for example this has 4 byte "hole" between the first two LOAD segments:

```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x001000 0x00000000 0x00000000 0x04834 0x04834 R E 0x1000
  LOAD           0x005838 0x00004838 0x00004838 0x01620 0x01620 R   0x1000
```

I'll write `emit_load_segments()` a bit differently.

Reply:
OK, some padding for proper alignment I suppose.